url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/function/excel/COUNTIF.java

```diff
@@ -0,0 +1,30 @@
+package org.boris.expr.function.excel;
+
+import org.boris.expr.Expr;
+import org.boris.expr.ExprArray;
+import org.boris.expr.ExprDouble;
+import org.boris.expr.ExprError;
+import org.boris.expr.ExprException;
+import org.boris.expr.IEvaluationContext;
+import org.boris.expr.function.AbstractFunction;
+import org.boris.expr.util.Condition;
+
+public class COUNTIF extends AbstractFunction
+{
+    public Expr evaluate(IEvaluationContext context, Expr[] args)
+            throws ExprException {
+        assertArgCount(args, 2);
+        // TODO assert first argument as reference
+        ExprArray array = asArray(context, args[0], true);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;context&#39; (index: 0 | type: IEvaluationContext). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;args&#39;. The same parameter type is validated here for example: url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/AbstractComparisonOperator.java:19. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).